code/modules/client/client_defines.dm

```diff
@@ -53,3 +53,31 @@
 
 	//So admins know why it isn't working - Used to determine what other accounts previously logged in from this computer id
 	var/related_accounts_cid = "Requires database"
+
+	///Amount of keydowns in the last keysend checking interval
+	var/client_keysend_amount = 0
+	///World tick time where client_keysend_amount will reset
+	var/next_keysend_reset = 0
+	///World tick time where keysend_tripped will reset back to false
+	var/next_keysend_trip_reset = 0
+	///When set to true, user will be autokicked if they trip the keysends in a second limit again
+	var/keysend_tripped = FALSE
+	///custom movement keys for this client
+	var/list/movement_keys = list()
+	///Are we locking our movement input?
+	var/movement_locked = FALSE
+
+	/// A buffer of currently held keys.
+	var/list/keys_held = list()
+	/// A buffer for combinations such of modifiers + keys (ex: CtrlD, AltE, ShiftT). Format: `"key"` -> `"combo"` (ex: `"D"` -> `"CtrlD"`)
+	var/list/key_combos_held = list()
+	/*
+	** These next two vars are to apply movement for keypresses and releases made while move delayed.
+	** Because discarding that input makes the game less responsive.
+	*/
+	/// On next move, add this dir to the move that would otherwise be done
+	var/next_move_dir_add
+	/// On next move, subtract this dir from the move that would otherwise be done
+	var/next_move_dir_sub
```

Comment:
To summarize several previous comments: I just don't understand what this does which `keys_held` isn't already doing. They don't seem to hold any additional information.